controller/AdminAccessController.php

```diff
@@ -124,36 +111,96 @@ public function savePermissions()
             ), 500);
         }
 
-        //get resource
-        $clazz = new \core_kernel_classes_Class($resourceId);
         $resources = array($clazz);
         if($recursive){
             $resources = array_merge($resources, $clazz->getSubClasses(true));
             $resources = array_merge($resources, $clazz->getInstances(true));
         }
 
-        foreach($resources as $resource){
-            $permissions = $this->dataAccess->getDeltaPermissions($resource->getUri(),$privileges);
-            // add permissions
-            foreach ($permissions['add'] as $userId => $privilegeIds) {
-                if(count($privilegeIds) > 0){
-                    $this->dataAccess->addPermissions($userId, $resource->getUri(), $privilegeIds);
-                }
+        $success = true;
+        $code = 200;
+        $message = __('Permissions saved');
+        $processedResources = [];
+        foreach ($resources as $resource) {
+            $permissions = $this->dataAccess->getDeltaPermissions($resource->getUri(), $privileges);
+            try {
+                $this->removePermissions($permissions['remove'], $resource);
+            } catch (\common_exception_InconsistentData $e) {
+                $success = false;
+                $message = $e->getMessage();
+                break;

```

Comment:
try catch could be outside of the foreach and you don't need to break the loop just handle the error outside. no success check just rollback in catch